File: src/1498.py
import sys
import logging
from typing import List, Optional
from collections import defaultdict, deque, Counter
from functools import cache, lru_cache
from bisect import bisect_left, bisect_right
from itertools import accumulate
from heapq import heapify, heappush, heappop


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def __init__(self):
        self.MODULO = 10 ** 9 + 7

    def numSubseq(self, nums: List[int], target: int) -> int:
        retVal = 0

        nums.sort()

        head = 0
        tail = len(nums) - 1
        while head <= tail:
            if nums[head] + nums[tail] > target:
                tail -= 1
                continue
            retVal += pow(2, (tail - head))
            retVal %= self.MODULO
            head += 1

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for nums, target in zip([[3, 5, 6, 7], [3, 3, 6, 8], [2, 3, 3, 4, 6, 7]], [9, 10, 12]):
            # /* Example
            #  *  Input: nums = [3,5,6,7], target = 9
            #  *  Output: 4
            #  *
            #  *  Input: nums = [3,3,6,8], target = 10
            #  *  Output: 6
            #  *
            #  *  Input: nums = [2,3,3,4,6,7], target = 12
            #  *  Output: 61
            #  */
            logging.info("Input: nums = %s, target = %s", nums, target)

            retVal = pSolution.numSubseq(nums, target)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass

    sys.exit(0)

File: src/1498.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
#define MODULO (int)(1e9 + 7)
   public:
    int numSubseq(vector<int>& nums, int target) {
        int retVal = 0;

        int numsSize = nums.size();
        vector<int> pows(numsSize, 1);
        for (int i = 1; i < numsSize; ++i) {
            pows[i] = pows[i - 1] * 2 % MODULO;
        }

        sort(nums.begin(), nums.end());
        int head = 0;
        int tail = numsSize - 1;
        while (head <= tail) {
            if (nums[head] + nums[tail] > target) {
                --tail;
                continue;
            }

            retVal += pows[tail - head];
            retVal %= MODULO;
            ++head;
        }

        return retVal;
    }
};

int main(int argc, char** argv) {
    struct subject {
        vector<int> nums;
        int target;
    };
    vector<subject> testData{{{3, 5, 6, 7}, 9}, {{3, 3, 6, 8}, 10}, {{2, 3, 3, 4, 6, 7}, 12}};
    int numberOfTestCase = testData.size();
    /* Example
     *  Input: nums = [3,5,6,7], target = 9
     *  Output: 4
     *
     *  Input: nums = [3,3,6,8], target = 10
     *  Output: 6
     *
     *  Input: nums = [2,3,3,4,6,7], target = 12
     *  Output: 61
     */

    Solution cSolution;
    int answer;
    for (int i = 0; i < numberOfTestCase; ++i) {
        cout << "Input: nums = [";
        for (long unsigned int j = 0; j < testData[i].nums.size(); ++j) {
            cout << ((j == 0) ? "" : ",") << testData[i].nums[j];
        }
        cout << "], target = " << testData[i].target << "\n";

        answer = cSolution.numSubseq(testData[i].nums, testData[i].target);
        cout << "Output: " << answer << "\n";

        cout << "\n";
    }

    return EXIT_SUCCESS;
}
